### PR TITLE
New way to config Firestore SDK Cache.

### PR DESCRIPTION
**Pull request by wu-hui**
Manual test link (private repo): https://github.com/FirebasePrivate/firestore-team/pull/50

### CODE REVIEW

File: packages/firestore/src/api/cache_config.ts [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+class IndexedDbLocalCacheImpl implements IndexedDbLocalCache {
+  kind: 'indexeddb' = 'indexeddb';
+  _onlineComponentProvider: OnlineComponentProvider;
+  _offlineComponentProvider: OfflineComponentProvider;
+
+  constructor(settings: IndexedDbSettings | undefined) {
+    let tabManager: IndexedDbTabManager;
+    if (settings?.tabManager) {
+      settings.tabManager.initialize(settings);
```
Comment (MarkDuckworth) [inline packages/firestore/src/api/cache_config.ts:57]: I haven't read through everything yet, but flagging this to revisit. It's a little strange that this constructor initializes the `settings.tabManager` with the `settings` argument. This seems like it falls outside the responsibility of this class. It might feel better if this block checked if the tabManager was not already initialized.

File: packages/firestore/src/api/cache_config.ts [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+export type FirestoreLocalCache = MemoryLocalCache | IndexedDbLocalCache;
+
+// Factory function
+export function memoryLocalCache(): MemoryLocalCache {
+  return new MemoryLocalCacheImpl();
+}
+
+export interface IndexedDbSettings {
+  cacheSizeBytes?: number;
```
Comment (MarkDuckworth) [inline packages/firestore/src/api/cache_config.ts:76]: Is there any reason to offer cache size config for memory local cache?

File: packages/firestore/src/api/cache_config.ts [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+export function indexedDbLocalCache(
+  settings?: IndexedDbSettings
+): IndexedDbLocalCache {
+  return new IndexedDbLocalCacheImpl(settings);
+}
+
+export interface IndexedDbSingleTabManager {
+  kind: 'indexedDbSingleTab';
+  initialize: (
+    settings: Omit<IndexedDbSettings, 'tabManager'> | undefined
```
Comment (MarkDuckworth) [inline packages/firestore/src/api/cache_config.ts:91]: I'm curious how using `Omit<...,...>` will turn out in the documentation. IMO, some of our documentation of settings objects can be hard to find or offers limited information. Not sure if this will make the problem worse. We could just define a new settings interface for this case.

File: packages/firestore/src/api/cache_config.ts (hunk 2 above)
Comment (MarkDuckworth) [reply packages/firestore/src/api/cache_config.ts:76]: Looks like we already offer cacheSize config for memory persistence in the lite SDK. Consider moving that setting here.

File: packages/firestore/src/core/firestore_client.ts [hunk 4]
```diff
@@ -120,12 +124,12 @@ export class FirestoreClient {
...
-     * (incoming GRPC messages), we should always schedule onto this queue.
-     * This ensures all of our work is properly serialized (e.g. we don't
-     * start processing a new operation while the previous one is waiting for
-     * an async I/O to complete).
+     * Asynchronous queue responsible for all of our internal processing. When //
+     * we get incoming work from the user (via public API) or the network //
+     * (incoming GRPC messages), we should always schedule onto this queue. //
+     * This ensures all of our work is properly serialized (e.g. we don't //
+     * start processing a new operation while the previous one is waiting for //
+     * an async I/O to complete). //
```
Comment (MarkDuckworth) [inline packages/firestore/src/core/firestore_client.ts:132]: Was this change intentional?

**Review by MarkDuckworth — APPROVED** [review]
I like the new interface! I added a few comments for your consideration.

File: packages/firestore/src/api/cache_config.ts (hunk 3 above)
Comment (wu-hui) [reply packages/firestore/src/api/cache_config.ts:91]: My intention is to hide this from public API. Renamed it to `_initialize`

File: packages/firestore/src/api/cache_config.ts (hunk 2 above)
Comment (wu-hui) [reply packages/firestore/src/api/cache_config.ts:76]: Once we introduce memory LRU GC, then we will have a cache size. For eager GC though, size has no effect.

File: packages/firestore/src/api/cache_config.ts (hunk 1 above)
Comment (wu-hui) [reply packages/firestore/src/api/cache_config.ts:57]: I agree with you. I could not come up with a better way however, because the way the underlying component providers need `cacheSizeBytes`, which comes from `setting` along size `tabManager`. Was hoping I can use your help to come up with a better way.

File: packages/firestore/src/api/index_configuration.ts [hunk 5]
```diff
@@ -151,17 +150,17 @@ export function setIndexConfiguration(
...
-
-  // PORTING NOTE: We don't return an error if the user has not enabled
-  // persistence since `enableIndexeddbPersistence()` can fail on the Web.
-  if (!client.offlineComponents?.indexBackfillerScheduler) {
+  if (
+    !client._uninitializedComponentsProvider ||
+    client._uninitializedComponentsProvider?._offlineKind === 'memory'
+  ) {
+    // PORTING NOTE: We don't return an error if the user has not enabled
+    // persistence since `enableIndexeddbPersistence()` can fail on the Web.
```
Comment (MarkDuckworth) [inline packages/firestore/src/api/index_configuration.ts:158]: This comment may need to be updated for future readers.

File: packages/firestore/src/api/database.ts [hunk 6]
```diff
@@ -287,14 +295,20 @@ export function configureFirestore(firestore: Firestore): void {
...
+    firestore._firestoreClient._uninitializedComponentsProvider = {
+      _offlineKind: settings.cache.kind,
+      _offline: settings.cache._offlineComponentProvider,
+      _online: settings.cache._onlineComponentProvider
     };
   }
 }
 
 /**
+ * NOTE: this function will be deprecated in a future major release. Instead, set
```
Comment (MarkDuckworth) [inline packages/firestore/src/api/database.ts:307]: Consider using the `@deprecated` tag. https://api-extractor.com/pages/tsdoc/tag_deprecated/

File: packages/firestore/src/api/cache_config.ts [hunk 7]
```diff
@@ -23,86 +23,185 @@ import {
...
+  toJSON(): {} {
+    return { kind: this.kind };
+  }
 }
 
-export interface IndexedDbLocalCache {
+/**
+ * Provides a cache backed by IndexedDb to the SDK.
+ *
+ * To use, create an instance using the factory function `indexedDbLocalCache()`, then
```
Comment (MarkDuckworth) [inline packages/firestore/src/api/cache_config.ts:71]: Consider using `{@link indexDbLocalCache()}` to generate a link in documentation to this factory. Also can apply above to memory cache.

**Review by MarkDuckworth — COMMENTED** [review]
LGTM. A few comments given for your consideration.

**Comment by wu-hui** [comment]
@markarndt please review the comments in:
packages/firestore/src/api/cache_config.ts
packages/firestore/src/api/database.ts

Thanks!

File: packages/firestore/src/api/cache_config.ts [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+  IndexedDbOfflineComponentProvider,
+  MemoryOfflineComponentProvider,
+  MultiTabOfflineComponentProvider,
+  OfflineComponentProvider,
+  OnlineComponentProvider
+} from '../core/component_provider';
+
+/* eslint @typescript-eslint/consistent-type-definitions: ["error", "type"] */
+/**
+ * Provides a in-memory cache to the SDK. This is the default cache unless explicitly
```
Comment (markarndt) [inline packages/firestore/src/api/cache_config.ts:28]: "...an in-memory..."

File: packages/firestore/src/api/cache_config.ts [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+ * specified otherwise.
```
Comment (markarndt) [inline packages/firestore/src/api/cache_config.ts:29]: configured

File: packages/firestore/src/api/cache_config.ts [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+}
+
+/**
+ * Union type from all supported SDK cache layer.
+ */
+export type FirestoreLocalCache = MemoryLocalCache | IndexedDbLocalCache;
+
+/**
+ * Creates an instance of `MemoryLocalCache`. The instance can be set to
+ * `FirestoreSettings.cache` to tell the SDK what cache layer to use.
```
Comment (markarndt) [inline packages/firestore/src/api/cache_config.ts:123]: which cache layer

File: packages/firestore/src/api/cache_config.ts [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+}
+
+/**
+ * An settings object to configure an `IndexedDbLocalCache` instance.
+ */
+export type IndexedDbCacheSettings = {
+  /**
+   * An approximate cache size threshold for the on-disk data. If the cache
+   * grows beyond this size, Firestore will start removing data that hasn't been
+   * recently used. The size is not a guarantee that the cache will stay below
```
Comment (markarndt) [inline packages/firestore/src/api/cache_config.ts:136]: The SDK does not guarantee that the cache...

File: packages/firestore/src/api/cache_config.ts [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+
+  /**
+   * Specifies how multiple tabs/windows will be managed by the SDK.
+   */
+  tabManager?: IndexedDbTabManager;
+};
+
+/**
+ * Creates an instance of `IndexedDbLocalCache`. The instance can be set to
+ * `FirestoreSettings.cache` to tell the SDK what cache layer to use.
```
Comment (markarndt) [inline packages/firestore/src/api/cache_config.ts:153]: which cache layer

File: packages/firestore/src/api/cache_config.ts [hunk 13]
```diff
@@ -0,0 +1,302 @@
...
+export type IndexedDbTabManager =
+  | IndexedDbSingleTabManager
+  | IndexedDbMultipleTabManager;
+
+/**
+ * Type to configure an `IndexedDbSingleTabManager` instace.
+ */
+export type IndexedDbSingleTabManagerSettings = {
+  /**
+   * Whether to force enable indexeddb cache for the client. This cannot be used
```
Comment (markarndt) [inline packages/firestore/src/api/cache_config.ts:279]: "force-enable IndexedDB"

Note hyphen and capitalization of IndexedDB.... Apply capitalization here and in other occurrences in this paragraph.

**Review by markarndt — APPROVED** [review]

File: packages/firestore/src/api/cache_config.ts (hunk 8 above)
Comment (wu-hui) [reply packages/firestore/src/api/cache_config.ts:28]: Done.

File: packages/firestore/src/api/cache_config.ts (hunk 9 above)
Comment (wu-hui) [reply packages/firestore/src/api/cache_config.ts:29]: Done.

File: packages/firestore/src/api/cache_config.ts (hunk 10 above)
Comment (wu-hui) [reply packages/firestore/src/api/cache_config.ts:123]: Done.

File: packages/firestore/src/api/cache_config.ts (hunk 11 above)
Comment (wu-hui) [reply packages/firestore/src/api/cache_config.ts:136]: Done.

File: packages/firestore/src/api/cache_config.ts (hunk 12 above)
Comment (wu-hui) [reply packages/firestore/src/api/cache_config.ts:153]: Done.

File: packages/firestore/src/api/cache_config.ts (hunk 13 above)
Comment (wu-hui) [reply packages/firestore/src/api/cache_config.ts:279]: Done.

File: packages/firestore/src/api/database.ts [hunk 14]
```diff
@@ -283,9 +290,24 @@ export function configureFirestore(firestore: Firestore): void {
...
+    firestore._firestoreClient._uninitializedComponentsProvider = {
+      _offlineKind: settings.cache.kind,
+      _offline: settings.cache._offlineComponentProvider,
+      _online: settings.cache._onlineComponentProvider
+    };
+  }
 }
 
 /**
+ * @deprecated This function will be removed in a future major release. Instead, set
```
Comment (hsubox76) [inline packages/firestore/src/api/database.ts:306]: I'm not sure but I think this tag needs to go at the bottom of the comment block. You can double check when you run yarn docs devsite.

**Review by hsubox76 — COMMENTED** [review]
Can you run `yarn docgen devsite` to generate new files in docs-devsite? There's a message in https://github.com/firebase/firebase-js-sdk/actions/runs/4345091205/jobs/7589318276 but it's not very easy to find, sorry.

File: packages/firestore/src/api/database.ts (hunk 14 above)
Comment (wu-hui) [reply packages/firestore/src/api/database.ts]: Done in the latest commit.

File: .changeset/brown-beers-tease.md [hunk 15]
```diff
@@ -0,0 +1,6 @@
+---
+"@firebase/firestore": minor
+"firebase-repo-scripts-prune-dts": minor
```
Comment (hsubox76) [inline .changeset/brown-beers-tease.md:3]: This won't work, this package is ignored by changesets. Also I don't think there's any changes to it in this PR?

Comment (wu-hui) [reply .changeset/brown-beers-tease.md]: Fixed. I modified some scripts at one point..

**Review by hsubox76 — APPROVED** [review]
